src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs

```diff
@@ -259,6 +260,27 @@ public override ImmutableArray<Location> Locations
             }
         }
 
+        /// <summary>
+        /// Locations[0] on lambda symbols covers the entire syntax, which is inconvenient but remains for compatibility.
+        /// For better diagnostics quality, use the ErrorLocation instead, which points to the "delegate" or the "=>".
+        /// </summary>
+        internal override Location ErrorLocation
+        {
+            get
+            {
+                switch (_syntax.Kind())
+                {
+                    case SyntaxKind.AnonymousMethodExpression:
+                        return ((AnonymousMethodExpressionSyntax)_syntax).DelegateKeyword.GetLocation();
+                    case SyntaxKind.SimpleLambdaExpression:
+                    case SyntaxKind.ParenthesizedLambdaExpression:
+                        return ((LambdaExpressionSyntax)_syntax).ArrowToken.GetLocation();
+                    default:
```

Comment:
I fear this won't work with query expressions. Do you have a test for that? Is that even possible? Even if we don't have a test exercising it today or a scenario where it can happen in the compiler as it stands today, this feels like a trap waiting to happen. It feels like it would be safer for the default case to return `Locations[0]`.